lib/IRGen/IRGenModule.cpp

```diff
@@ -1098,10 +1098,20 @@ void IRGenModule::addLinkLibrary(const LinkLibrary &linkLib) {
   
   switch (linkLib.getKind()) {
   case LibraryKind::Library: {
-    llvm::SmallString<32> opt =
-        getTargetDependentLibraryOption(Triple, linkLib.getName());
-    AutolinkEntries.push_back(
-        llvm::MDNode::get(ctx, llvm::MDString::get(ctx, opt)));
+    if (TargetInfo.OutputObjectFormat == llvm::Triple::ELF &&
+        IRGen.Opts.LLVMLTOKind != IRGenLLVMLTOKind::None) {
+      // When performing LTO, we always use lld that supports auto linking
+      // mechanism with ELF. So embed dependent libraries names in
+      // "llvm.dependent-libraries" instead of options to avoid using
+      // swift-autolink-extract.
+      AutolinkEntries.push_back(
```

Comment:
I really dislike that this is being done here in this way. It leaks an implementation detail. Why not change AutolinkEntries to instead of taking an MDNode, take a type called:

```
struct AutolinkEntry {
   PointerIntPair<llvm::MDNode *, 1> mdNodeOrLibName;
}
```

Then you can use the lib name as a discriminator and move this code to emitAutolinkInfo maybe in a helper function. Your thoughts?